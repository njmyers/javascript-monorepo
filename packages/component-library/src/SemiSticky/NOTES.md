# SemiSticky

SemiSticky is a position aware component that animates in and out on scroll positions. It is inspired by the CSS property `position: sticky` but allows for usage in a much wider variety of situations. SemiSticky uses the AnimationProps for determining it's style. Please see AnimationProps for more information about usage.

SemiSticky uses a single prop to determine the scroll position of it's on state. That prop is called `top` and it signifies the amount of pixels from the top of the page that the component should apply it's onState styles. Anything greater then top will apply the onState styles and anything less then top will apply the offState styles.

## Usage

A common usage pattern is for fixed position headers and footers that will show themselves based on a user's scroll position.

```js
import React from 'react';

class Main extends React.Component {
  render() {
    return (
      <SemiSticky top={400}>
        <Header />
      </SemiSticky>
    );
  }
}
```

In the above example, the Header component will hide from view when the user scrolls down 400 pixels. Sounds simple enough but to implement yourself takes many lines of code! Feel free to add your own onState and offState styles and SemiSticky will automatically generate CSS transitions.

If you would like to shallowly merge styles, use the `style` prop. If you want to replace the default inline styles, use the `replaceStyle` prop. You can also apply `className` prop but keep in mind that all of the inline styles will take precedence.

<!-- STORY -->
