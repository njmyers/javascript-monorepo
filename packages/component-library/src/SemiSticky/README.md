# SemiSticky

SemiSticky is a position aware component that animates in and out on scroll positions. It is inspired by the CSS property `position: sticky` but allows for usage in a much wider variety of situations. SemiSticky uses the AnimationProps for determining it's style. Please see AnimationProps for more information about usage.

SemiSticky uses a single prop to determine the scroll position of it's on state. That prop is called `top` and it signifies the amount of pixels from the top of the page that the component should apply it's onState styles. Anything greater then top will apply the onState styles and anything less then top will apply the offState styles.

## Usage

A common usage pattern is for fixed position headers and footers that will show themselves based on a user's scroll position.

```js
import React from 'react';

class Main extends React.Component {
  render() {
    return (
      <SemiSticky top={400}>
        <Header />
      </SemiSticky>
    );
  }
}
```

In the above example, the Header component will hide from view when the user scrolls down 400 pixels. Sounds simple enough but to implement yourself takes many lines of code! Feel free to add your own onState and offState styles and SemiSticky will automatically generate CSS transitions.

If you would like to shallowly merge styles, use the `style` prop. If you want to replace the default inline styles, use the `replaceStyle` prop. You can also apply `className` prop but keep in mind that all of the inline styles will take precedence.

<!-- STORY -->

## Props
| Prop Name | Type | Required | Default Value | Description |
|:----------|:-----|:---------|:--------------|:------------|
|`children`|`ReactNode`|false|-|_react children (your component)_|
|`className`|`string`|false|-|_className applied to the container element_|
|`offState`|<code>{}</code>|false|<code>{<br>  transform: 'translateY(-100px)',<br>}</code>|_css inline styles applied to the off state_|
|`onState`|<code>{}</code>|false|<code>{<br>  transform: 'translateY(0)',<br>}</code>|_css inline styles applied to the on state_|
|`replaceStyle`|<code>{}</code>|false|<code>{<br>  position: 'fixed',<br>  background: 'rgba(0, 0, 0, 0.8)',<br>  width: '100%',<br>  top: 0,<br>  left: 0,<br>}</code>|_completely replace all styles_|
|`sizes`|<code>{<br>  isSemiStickyActive: boolean,<br>}</code>|true|-|_inherited sizing info_|
|`style`|<code>{}</code>|false|-|_shallowly merge styles_|
|`top`|`number`|false|<code>200</code>|_distance from the document top to engage the on state styles_|
|`transitionSpeed`|`number`|false|<code>0.25</code>|_the speed of the transition_|
|`transitionTiming`|<code>&#124; 'ease'<br>&#124; 'linear'<br>&#124; 'ease-in'<br>&#124; 'ease-out'<br>&#124; 'ease-in-out'<br>&#124; 'step-start'<br>&#124; 'step-end'</code>|false|<code>'ease'</code>|_the transition timing function_|