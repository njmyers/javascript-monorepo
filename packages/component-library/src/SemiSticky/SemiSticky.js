/// @flow\
import * as React from 'react';
import withSize from 'react-size-components';

type Props = {
  sizes: {
    isSemiStickyActive: boolean,
  },
  children?: React.Node,
  style: {},
  replaceStyle: {},
  className: string,
  onState: {},
  offState: {},
  transitionSpeed: number,
  transitionTiming:
    | 'ease'
    | 'linear'
    | 'ease-in'
    | 'ease-out'
    | 'ease-in-out'
    | 'step-start'
    | 'step-end',
};

type State = {
  style: {},
  isSemiStickyActive: boolean,
};

// TODO: Add styles like this ...style ...onState, ...offState, ...replaceStyle

/**
 * Makes a semi sticky component
 * Creates an aside HTML5 element and adds your content inside of it
 * @param {object} style define the styles you want to add to the aside (shallow merge)
 */
class SemiSticky extends React.Component<Props, State> {
  static defaultProps = {
    replaceStyle: {
      position: 'fixed',
      background: 'rgba(0, 0, 0, 0.8)',
      width: '100%',
      top: 0,
      left: 0,
    },
    onState: {
      transform: 'translateY(0)',
    },
    offState: {
      transform: 'translateY(-100px)',
    },
    transitionTiming: 'ease',
    transitionSpeed: 0.25,
  };

  getTransitionString = () => {
    const { onState } = this.props;
    const { transitionSpeed, transitionTiming } = this.props;

    return Object.keys(onState)
      .map((key: string) => `${transitionSpeed}s ${transitionTiming} ${key}`)
      .reduce((a, b) => `${a},${b}`);
  };

  transition = this.getTransitionString();

  style = () => ({
    transition: this.transition,
    ...(this.props.sizes.isSemiStickyActive
      ? this.props.onState
      : this.props.offState),
    ...this.props.replaceStyle,
    ...this.props.style,
  });

  render() {
    return (
      <React.Fragment>
        <div />
        <aside className={this.props.className} style={this.style()}>
          {this.props.children}
        </aside>
      </React.Fragment>
    );
  }
}

type WrapperProps = {
  top: number,
};

/**
 * Essentially just turning a HOC into a child type usage.
 * Extremely Verbose here
 * Need to rewrite size components so you can use it in HOC or other ways
 * @extends React
 */
class HOCWrapper extends React.Component<WrapperProps> {
  static defaultProps = {
    top: 480,
  };

  custom = [
    {
      name: 'isSemiStickyActive',
      fn: (node: HTMLElement) => {
        const top = Math.abs(node.getBoundingClientRect().top);
        return top > this.props.top;
      },
      subscriptions: {
        scroll: true,
        resize: true,
      },
      schema: '',
    },
  ];

  render() {
    const Wrapper = withSize({ custom: this.custom })(SemiSticky);
    return <Wrapper {...this.props} />;
  }
}
export default HOCWrapper;
