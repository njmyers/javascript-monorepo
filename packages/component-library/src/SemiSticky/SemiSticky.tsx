// @ts-nocheck
import * as React from 'react';
import withSize from 'react-size-components';
import createTransitionStyle from '../utils/create-transition-style';

type AnimationProps = {
  /** className applied to the container element */
  className?: string,
  /** css inline styles applied to the on state */
  onState: {},
  /** css inline styles applied to the off state */
  offState: {},
  /** completely replace all styles */
  replaceStyle: {},
  /** shallowly merge styles */
  style?: {},
  /** the speed of the transition */
  transitionSpeed: number,
  /** the transition timing function */
  transitionTiming:
    | 'ease'
    | 'linear'
    | 'ease-in'
    | 'ease-out'
    | 'ease-in-out'
    | 'step-start'
    | 'step-end',
};

type InheritedProps = {
  /** inherited sizing info */
  sizes: {
    isSemiStickyActive: boolean,
  },
  /** reference from withSize for dom component position/sizing */
  childRef?: HTMLElement,
  /** react children (your component) */
  children?: React.Node,
};

/**
 * Merges the styles according to the boolean flag
 * Creates an aside HTML5 element and adds your content inside of it
 * Also creates an empty div so we have somewhere to measure from
 */
class StyleMerger extends React.Component<InheritedProps & AnimationProps> {
  style = () => ({
    ...createTransitionStyle(this.props),
    ...(this.props.sizes.isSemiStickyActive
      ? this.props.onState
      : this.props.offState),
    ...this.props.replaceStyle,
    ...this.props.style,
  });

  render() {
    return (
      <React.Fragment>
        <div ref={this.props.childRef} />
        <aside className={this.props.className} style={this.style()}>
          {this.props.children}
        </aside>
      </React.Fragment>
    );
  }
}

type WrapperProps = {
  /** distance from the document top to engage the on state styles */
  top: number,
};

type AllProps = WrapperProps & AnimationProps & InheritedProps;

/**
 * SemiSticky is a position aware component that animates in and out on scroll positions.
 * It is inspired by the CSS property `position: sticky` but allows for usage in a much wider variety of situations.
 * SemiSticky uses the AnimationProps for determining it's style. Please see AnimationProps for more information about usage.
 * SemiSticky uses a single prop to determine the scroll position of it's on state.
 * That prop is called `top` and it signifies the amount of pixels from the top of the page that the component should apply it's onState styles.
 * Anything greater then top will apply the onState styles and anything less then top will apply the offState styles.
 * A common usage pattern is for fixed position headers and footers that will show themselves based on a user's scroll position.
 */
class SemiSticky extends React.Component<AllProps> {
  static defaultProps = {
    top: 200,
    replaceStyle: {
      position: 'fixed',
      background: 'rgba(0, 0, 0, 0.8)',
      width: '100%',
      top: 0,
      left: 0,
    },
    onState: {
      transform: 'translateY(0)',
    },
    offState: {
      transform: 'translateY(-100px)',
    },
    transitionTiming: 'ease',
    transitionSpeed: 0.25,
  };

  static displayName = 'SemiSticky';

  custom = [
    {
      name: 'isSemiStickyActive',
      fn: (node: HTMLElement) => {
        const top = Math.abs(node.getBoundingClientRect().top);
        return top > this.props.top;
      },
      subscriptions: {
        scroll: true,
        resize: true,
      },
      schema: '',
    },
  ];

  render() {
    const Wrapper = withSize({ custom: this.custom })(StyleMerger);
    return <Wrapper {...this.props} />;
  }
}

export default SemiSticky;
