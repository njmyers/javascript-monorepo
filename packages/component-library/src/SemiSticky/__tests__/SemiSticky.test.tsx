import React from 'react';
import SemiSticky from '../SemiSticky';

import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('the input works', () => {
  test('the input mounts without crashing', () => {
    const wrapper = shallow(<SemiSticky />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
