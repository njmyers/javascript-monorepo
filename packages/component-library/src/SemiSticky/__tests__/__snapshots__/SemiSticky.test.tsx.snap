// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the input works the input mounts without crashing 1`] = `"<div></div><aside style=\\"transition:0.25s ease transform;transform:translateY(-100px);position:fixed;background:rgba(0, 0, 0, 0.8);width:100%;top:0;left:0\\"></aside>"`;
